File: implementation1/utilities.m
function plot_dtft(x)
    N = max(length(x), 100000);
    
    X = fft(x, N);
    X = fftshift(X);
    
    w = (-N/2:N/2-1) * (2*pi/N);
    
    figure;
    plot(w, abs(X));
    grid on;
    
    xlim([-pi pi]);
    xticks([-pi, -3*pi/4, -pi/2, -pi/4, 0, pi/4, pi/2, 3*pi/4, pi]);
    
    ylabel('|X(\omega)|');
    xlabel('\omega (rad/sample)');
end


% Example usage
h = [1, 2, 3, 2, 1];
plot_dtft(h);

File: implementation1/utilities.py
def plot_dtft(x):
    import numpy as np
    import matplotlib.pyplot as plt
    N = max(len(x), 100000)
    X = np.fft.fft(x, n=N)
    X = np.fft.fftshift(X)
    w = np.linspace(-np.pi, np.pi, N, endpoint=False)

    _, ax = plt.subplots(nrows=1, ncols=1, tight_layout=True)
    mag_ax = ax
    mag_ax.plot(w, np.abs(X))
    mag_ax.set_xticks(
        [-np.pi, -3*np.pi/4, -np.pi/2, -np.pi/4, 0, np.pi/4, np.pi/2, 3*np.pi/4, np.pi],
    )
    for x, label in [
        (-np.pi, r'$-\pi$'),
        (-np.pi/2, r'$-\pi/2$'),
        (np.pi/2, r'$\pi/2$'),
        (np.pi, r'$\pi$'),
    ]:
        mag_ax.annotate(
            label,
            xy=(x, 0),
            xycoords=('data', 'axes fraction'),
            xytext=(0, -25),
            textcoords='offset points',
            ha='center',
            va='top'
        )
    mag_ax.set_ylabel(r'|$X(\omega)$|')
    mag_ax.grid()
    mag_ax.set_xlim(-np.pi, np.pi)
    mag_ax.set_xlabel(r'$\omega$ (rad/sample)')

    plt.show()


if __name__ == '__main__':
    # Example usage
    h = [1, 2, 3, 2, 1]
    plot_dtft(h)
